feat(API,CLI) add support for searches with complex filters

Barbozafab opened the pull request:
## Description

- Created a new option for `ls` and `get` to perform searches with complex filters;

- Created new API endpoint `/api/objects/search`;

- Updated the CLI autocompleter with `cp`, `connect3d` and `disconnect3d` commands;

- Updated the layers:
  - Layers used to have an attribute `filters`, which was a `map[string]string` filled with multiple simple filters; when the layer was applied, an `AND` operation was done between all the simple filters.
  - Now, this attribute has been changed to `filter`, a single string containing a complex filter; 
  - The complex filter can be reassigned by doing `[slug]:filter=[new complex filter]`;
  - A new filter can be added to the current one, resulting in an `AND` operation between the old filter and the new one, by doing `[slug]:filter+=[new complex filter]`.

- Fixes #278 

## Type of change

- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

- Local test

helderbetiol commented on wiki/CLI-language.md:
````diff
@@ -428,6 +428,14 @@ By adding the filter tag=[tag_slug] to the ls, we can obtain the children that h
+
+Complex filters can be added to the ls command with the flag `-f`, composing complex boolean expressions with the operators `=`, `!=`, `<`, `<=`, `>`, `>=`, `&` and `|`.
+
+```
````
adding more examples would be nice, as well as mentioning the () in the text. And what about the get?

Barbozafab replied:
I've rewritten this section of the Wiki to better explain the functioning of the complex filter, and added a section about it to the `get` command.

helderbetiol commented on CLI/controllers/layer.go:
```diff
@@ -167,11 +168,16 @@ func (controller Controller) UpdateLayer(path string, attributeName string, valu
-			return fmt.Errorf("invalid filter format")
+		var filters map[string]any
+		reg := regexp.MustCompile(`^(\w(\w|\-)*)=[^!<>=&|]+$`)
+		if strings.HasPrefix(value.(string), "filter=") {
```
Isn't it filters+= the prefix to add filters?

Barbozafab replied:
It is! Line 170 (`case models.LayerFiltersAdd:`) checks for this. Line 173 checks if the user has specified `filter=` in the complex filter. 
e.g. `test:filters+=filter=category!=room`

helderbetiol commented on CLI/utils/util.go:
```diff
@@ -255,3 +257,101 @@ func ObjectAttr(obj map[string]any, attr string) (any, bool) {
+		if match := re.FindStringSubmatch(expression); match != nil {
+			switch match[1] {
+			case "startDate":
+				// if match[2] != "=" {
```
let's erase the comments

helderbetiol commented on CLI/controllers/commandController.go:
```diff
@@ -136,11 +136,17 @@ func (controller Controller) ObjectUrlGeneric(pathStr string, depth int, filters
+	endpoint := "/api/objects"
 	for key, value := range filters {
-		params.Set(key, value)
+		fmt.Println("k, v", key, value)
```
bye bye print?

Barbozafab replied:
Bye bye print :)

helderbetiol commented:
If someone forgets to put the path in the ls command, maybe we should give a better error:
```
/Physical> ls -f reservedColor=AAAAAA
Error: on GET /api/hierarchy-objects/reservedColor=AAAAAA/all?limit=1 : cannot unmarshal json : 
404 page not found
```

helderbetiol commented:
The classic (simple) filter could use wildcards (*) as bellow while the complex one cannot:
 ```
 admin@TenantName:/Physical> ls . name=site*
k, v name site*
siteA
siteB
siteD
admin@TenantName:/Physical> ls . -f name=site*
k, v filter name=site*
Error: Invalid filter expression
 ```

Barbozafab commented on API/controllers/entityController.go:
```diff
@@ -506,7 +507,141 @@ func HandleGenericObjects(w http.ResponseWriter, r *http.Request) {
+}
+
+// swagger:operation POST /api/objects/search Objects GetGenericObject
+// Get all objects from any entity. Return as a list.
```
I've updated the endpoint name on the doc and inserted info about the body of the request.

Barbozafab commented:
I've updated the parser so now it demands a path to be specified when performing an `ls` or `get with a complex filter, and I've also added support for wildcards on complex filters.

helderbetiol commented on API/controllers/entityController.go:
```diff
@@ -506,7 +508,130 @@ func HandleGenericObjects(w http.ResponseWriter, r *http.Request) {
+// swagger:operation POST /api/objects/search Objects HandleComplexFilters
+// Get all objects from any entity that match the complex filter. Return as a list.
+// Wildcards can be used on any of the parameters present in query with equality and inequality operations.
+//
```
maybe just remove the explication of wildcard from here and mention "check endpoint X for info on wildcards"

Barbozafab replied:
Done!

helderbetiol commented on API/controllers/entityController.go:
```diff
@@ -506,7 +508,130 @@ func HandleGenericObjects(w http.ResponseWriter, r *http.Request) {
+			return
+		}
+
+		// Save entity to help delete and respond
```
Is this copied code useful here?

Barbozafab replied:
It is now that the `DELETE` method has been added to the `HandleComplexFilters` endpoint.

helderbetiol commented on API/models/model.go:
```diff
@@ -276,6 +280,91 @@ func GetManyObjects(entityStr string, req bson.M, filters u.RequestFilters, user
+	var err error
+	var c *mongo.Cursor
+
+	var opts *options.FindOptions
```
I think I've seen this code before, maybe we can create a function for it?
